File: saga-unity/Assets/Scripts/RPG/Effects/EffectStatus.cs
using System.Threading.Tasks;

public class EffectStatus : EffectEnemyTarget {

    protected new EffectStatusData data;

    private Status status;

    public EffectStatus(EffectStatusData data, CombatItem item) : base(data, item) {
        this.data = data;
        status = new Status(data.status);
    }

    protected override async Task ApplyEffect(Intent intent, Unit target, int power) {
        await target.TryApplyStatusAsync(status, intent.Battle);
    }

    protected override Task<int> CalculateAttackPowerAsync(Intent intent) {
        var power = data.hit;
        if (data.accStat != StatTag.NONE) {
            power += intent.Actor[data.accStat];
        }
        return Task.FromResult(power);
    }

    protected override async Task<bool> CheckHitAsync(Intent intent, Unit target, int power, float roll) {
        var tab = BattleBox.Tab;
        var temp = power;

        if (target.IsResistantTo(status)) {
            temp -= 255;
        }
        if (data.evadeStat != StatTag.NONE) {
            temp -= target[data.evadeStat];
        }
        float chance = temp / 100f;
        var hit = roll < chance;
        if (hit) {
            return true;
        } else {
            await intent.Battle.WriteLineAsync(tab + "Nothing happens.");
            return false;
        }
    }
}

File: src/saga-desktop/src/net/wombatrpgs/saga/rpg/warheads/EffectStatus.java
/**
 *  EffectStatus.java
 *  Created on Apr 28, 2014 7:09:32 PM for project saga-desktop
 *  Author: psy_wombats
 *  Contact: psy_wombats@wombatrpgs.net
 */
package net.wombatrpgs.saga.rpg.warheads;

import net.wombatrpgs.saga.core.SConstants;
import net.wombatrpgs.saga.rpg.battle.Battle;
import net.wombatrpgs.saga.rpg.battle.Intent;
import net.wombatrpgs.saga.rpg.chara.Chara;
import net.wombatrpgs.saga.rpg.chara.Status;
import net.wombatrpgs.saga.rpg.items.CombatItem;
import net.wombatrpgs.sagaschema.rpg.warheads.EffectStatusMDO;

/**
 * Inflicts status conditions on enemies.
 */
public class EffectStatus extends EffectEnemyTarget {
	
	protected EffectStatusMDO mdo;
	protected Status status;

	/**
	 * Inherited constructor.
	 * @param	mdo				The data to construct from
	 * @param	item			The item to construct for
	 */
	public EffectStatus(EffectStatusMDO mdo, CombatItem item) {
		super(mdo, item);
		this.mdo = mdo;
		status = Status.get(mdo.status);
	}

	/**
	 * @see net.wombatrpgs.saga.rpg.warheads.EffectEnemyTarget#onAffect
	 * (net.wombatrpgs.saga.rpg.battle.Battle, net.wombatrpgs.saga.rpg.battle.Intent,
	 * net.wombatrpgs.saga.rpg.chara.Chara, net.wombatrpgs.saga.rpg.chara.Chara, int)
	 */
	@Override
	protected void onAffect(Battle battle, Intent intent, Chara user, Chara victim, int power) {
		status.inflict(battle, victim);
	}

	/**
	 * @see net.wombatrpgs.saga.rpg.warheads.EffectEnemyTarget#calcPower
	 * (net.wombatrpgs.saga.rpg.battle.Battle, net.wombatrpgs.saga.rpg.chara.Chara)
	 */
	@Override
	protected int calcPower(Battle battle, Chara user) {
		int power = mdo.hit;
		if (mdo.accStat != null) {
			power += user.get(mdo.accStat);
		}
		return power;
	}

	/**
	 * @see net.wombatrpgs.saga.rpg.warheads.EffectEnemyTarget#hits
	 * (net.wombatrpgs.saga.rpg.battle.Battle, net.wombatrpgs.saga.rpg.chara.Chara,
	 * net.wombatrpgs.saga.rpg.chara.Chara, int, float)
	 */
	@Override
	protected boolean hits(Battle battle, Chara user, Chara target, int power, float roll) {
		String tab = SConstants.TAB;
		int temp = power;
		if (target.getStatus() != null) {
			if (target.getStatus().getPriority() > status.getPriority()) {
				temp -= 255;
			}
		}
		if (target.resists(status)) {
			temp -= 255;
		}
		if (mdo.evadeStat != null) {
			temp -= target.get(mdo.evadeStat);
		}
		float chance = (float) temp / 100f;
		boolean hit = roll < chance;
		if (hit) {
			return true;
		} else {
			battle.println(tab + "Nothing happens.");
			return false;
		}
	}

}
